### PR TITLE
Feature/add documentation

### PR DESCRIPTION
**Pull request by lleon95**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,111 @@
...
 
-# Dependencies
-1. meson 1.2.3
-2. Python 3.8.10
-3. C++17
+Framework to develop FPGA applications in C++ with the easiness of PYNQ
+
+## Introduction
+
+CYNQ is a C++ framework to implement FPGA-based accelerated applications with the same easiness as PYNQ framework for Python. This allows users to implement their own applications with better performance than in Python and avoids the long processing times of coding applications with Vitis.
```
Comment (dierpg) [inline README.md]: change easiness by ease of use

File: README.md [hunk 2]
```diff
@@ -1,7 +1,111 @@
...
+
+using namespace cynq;
+
+// Configure the FPGA
+auto kArch = HardwareArchitecture::UltraScale;
+auto platform = IHardware::Create(kArch, "design.bit", "default.xclbin");
+
+// Extract the accelerator (IP Core) and DMA
+// Addresses are given by the design
+auto accel = platform->GetAccelerator(0xa000000);
```
Comment (dierpg) [inline README.md]: can this be provided through a variable? looks like a magic number

File: README.md [hunk 3]
```diff
@@ -1,7 +1,111 @@
...
+auto dma = platform->GetDataMover(0xa0010000);
```
Comment (dierpg) [inline README.md]: same comment as before

File: README.md [hunk 4]
```diff
@@ -1,7 +1,111 @@
...
+## Links & References:
+
+* Docs: https://ecaslab.github.io/cynq
+* Github: https://github.com/ECASLab/cynq
+
+Cite Us:
+
+```
+@misc{cynq,
+  author = {{León-vega, Luis G.
```
Comment (dierpg) [inline README.md]: *Vega

File: docs/About.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# About
+
+CYNQ is an alternative framework to code applications for Xilinx FPGAs with the easiness of PYNQ. Currently, most of the flow is quite complex and requires Vitis to get a decent C++ application and there are no bindings from the official PYNQ. Based on these needs, CYNQ proposes:
```
Comment (dierpg) [inline docs/About.md:3]: replace easiness by pynq like api, the word keeps popping up, also change CYNQ proposes by "Cynq has the following objectives:"

File: docs/GettingStarted.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+6) To access the data from the memory buffers, you can use the `HostAddress` method.
+
+```c++
+using DataType = uint64_t;
+DataType* A = in_mem->HostAddress<DataType>().get();
+
+A[5] = 1;
+```
+
+The `HostAddress<T>()` maps the memory into a pointer that is accessible to the host. `T` can be any type that can be reinterpretedly casted.
```
Comment (dierpg) [inline docs/GettingStarted.md:72]: any type that supports reinterpret_pointer_cast and provide link to the cpp reference doc this is meant for newbies.
https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast

**Review by dierpg — CHANGES_REQUESTED** [review]
Pretty general changes on the redaction of the document, also a little comment on the use of cpp terminology that might not be obvious to new cpp users (which is part of the target audience).

File: README.md [hunk 7]
```diff
@@ -1,7 +1,111 @@
...
+inbuf = allocate(shape=(input_elements,), dtype=np.uint16)
+outbuf = allocate(shape=(output_elements,), dtype=np.uint16)
+
+# Run
+dma.sendchannel.transfer(inbuf)
+accel.write(accel.register_map.CTRL.address, 0x81)
+dma.recvchannel.transfer(outbuf)
+dma.recvchannel.wait()
+
+# Dispose the buffers
```
Comment (dierpg) [inline README.md:51]: Mention that this is handled with the RAII approach.

File: README.md [hunk 8]
```diff
@@ -1,7 +1,111 @@
...
+uint16_t* output_ptr = outbuf->HostAddress<uint16_t>().get();
+
+// Run
+accel->Start(StartMode::Continuous);
+inbuf->Sync(SyncType::HostToDevice);
+mover->Upload(in_mem, infbuf->Size(), 0, ExecutionType::Sync);
+mover->Download(out_mem, outbuf->Size(), 0, ExecutionType::Sync);
+outbuf->Sync(SyncType::DeviceToHost);
+
+// Dispose? We use RAII
```
Comment (dierpg) [inline README.md:87]: Provide a link with a resource that explains RAII.

File: docs/Foundations.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
+# Foundations
+
+## Current State
+
+To develop on Xilinx FPGAs, there are a couple of workflows:
+
+1. Legacy HLS -> Vivado workflow
```
Comment (dierpg) [inline docs/Foundations.md:7]: Maybe providing a section that goes through the useful resources for following the legacy HLS workflow might be handy, in a separate md file as an appendix.

File: docs/Foundations.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+2. Vitis Kernel workflow (newest)
```
Comment (dierpg) [inline docs/Foundations.md:8]: Maybe a comparison matrix can be made as a resource which can be useful if this information has to be consulted on a regular basis.

File: docs/GettingStarted.md [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+* `xclbin` is the path to the .xclbin file. You can use the one in `third-party/resources/default.xclbin`.
+
+3) Create the DMA instances to move the data. This is intended for designs that uses AXI4-Stream.
+
+```c++
+constexpr int kDmaAddress = 0xa0010000;
+auto dma = platform->GetDataMover(kDmaAddress);
+```
+
+where `kDmaAddress` is the address of the DMA instance we want to control. This is given by the design.
```
Comment (dierpg) [inline docs/GettingStarted.md:32]: Define what a design means in this context.

File: docs/GettingStarted.md [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+```c++
+constexpr int kAccelAddress = 0xa0000000;
+auto accel = platform->GetAccelerator(kAccelAddress);
+```
+
+where `kAccelAddress` is the AXI4-Lite control port of the IP core. As a requirement:
+
+> The IP Cores must have an AXI4-Lite port as a control port
+
+5) Get buffers to exchange data. These buffers are usually dual memory: they are mapped into host and device regions (physically contiguous).
```
Comment (dierpg) [inline docs/GettingStarted.md:45]: make examples of how these memory addresses look like, maybe making a diagram for this in the future can help illustrate this.

File: docs/GettingStarted.md [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+9) Transferring information requires the synchronisation of the memory buffers and the interaction with the data mover / DMA.
+
+To synchronise the buffers, it is possible to use `Sync()` method: 
+
+```c++
+in_mem->Sync(SyncType::HostToDevice);
+out_mem->Sync(SyncType::DeviceToHost);
+```
+
+that takes any of the following values:
```
Comment (dierpg) [inline docs/GettingStarted.md]: any can be interpreted as a list of values, clarify that only ONE value is passed to the method but it is ONE from the list provided.

**Review by dierpg — CHANGES_REQUESTED** [review]

File: docs/Foundations.md (hunk 9 above)
Comment (lleon95) [reply docs/Foundations.md:7]: Addressing it in other PR

File: docs/Foundations.md (hunk 10 above)
Comment (lleon95) [reply docs/Foundations.md:8]: Addressing it in other PR

File: docs/GettingStarted.md (hunk 12 above)
Comment (lleon95) [reply docs/GettingStarted.md:45]: Addressing it in other PR

**Pull request merged**